### ISSUE TITLE
[question] Groupwise training with group_size >=2

### ISSUE THREAD

**Issue by VittalP** [tensorflow/ranking#132]
I am trying to understand group_wise training better especially for group_size >= 2. Looking through the code, I see that the size of the tensor that is passed to`_score_fn()` is [`[batch_size * num_groups, self._group_size]`](https://github.com/tensorflow/ranking/blob/master/tensorflow_ranking/python/model.py#L373). So, if we consider a situation where each document has only one feature of `d` dimensions, does using a group_size = k, result in the input_layer being a `k*d` dimensional vector? Or, in the more general case, with context and group features, will the input_layer be `\sum_i cd_i + \sum_j k * gd_j` dimensional? Where `cd_i` is the size of the `i`-th context feature and `gd_j` is the size of the `j`-th group feature? If the answer to the above question is ‘yes’, does that mean that NN learns separate weights for the same feature across different documents of the group (because in the one feature case, the weight vector should now be `k*d` dimensional and not just `d`-dimensional)?

Secondly, I am confused how the output logits map to the appropriate input documents. From the example _score_fn() provide in this [tutorial](https://github.com/tensorflow/ranking/blob/master/tensorflow_ranking/examples/tf_ranking_libsvm.ipynb),  for a group_size = k, we need to create a final layer with `k` neuronal units in it.

```python
logits = tf.layers.dense(cur_layer, units=k)

```
 How does the model guarantee that the `i`-th logit corresponds to the `i`-th input document `x_i`?

In other words, my question is the following. If
```
[y1, y2,…y_k] = _score_fn([x1, x2, … x_k])
```
then for any permutation, `\pi`, applied on the input group, how can we guarantee that the output will also be permuted accordingly, i.e.

```
\pi([y1, y2,… y_k]) = _score_fn( \pi([x1, x2, … x_k]) )
```
?

**Comment by johnvorsten** [comment]
[Here is the paper](https://arxiv.org/abs/1811.04415) I found about groupwise ranking.  It helped me understand more of what groupwise ranking is and what group_size does.  

`will the input_layer be \sum_i cd_i + \sum_j k * gd_j dimensional?`
Yes, that is what I found when I looked at the computational graph

`separate weights for the same feature across different documents of the group`
That is the way I understand it.  For example, each context feature input would get a weight associated with each feature of each per-item example.  I only did complete list-wise rankings, in which each case a logit is assigned to an input document.  

`How does the model guarantee that the i-th logit corresponds to the i-th input document x_i`
I think that is how the scoring function works.  The scoring function will find the error for each output logit with the corresponding label.  Error is propagated back through the model and weights are updated.  The ith logit corresponds to the label of the ith input document.  I only did complete list-wise rankings (aka. group_size = list_size), so I didn't observe how it worked with when group_size < list_size.
Take a look at tfr.python.model._compute_logits_impl, and _update_scatter_gather_indices.  It shows how documents are formed into groups.  You will see how example features are gathered.  I think loss functions are in the ranking_head, so look there for how labels are gathered.  The paper I mentioned above gives motivation for ranking on groups.  It will probably answer your questions very well!

**Comment by xuanhuiwang** [comment]
Thanks for your answers, @johnvorsten. They look precise!

@VittalP, when you have group_size >= 2, there will be `k*d` weights, instead of `d` weights.

Indeed, `_update_scatter_gather_indices` stores the *original* document index and can be used to map the scores from `_score_fn` back to the original index for each document.

For methodology, you can read the paper @johnvorsten mentioned.

Closing the issues now. Feel free to reopen.

**Comment by korosig** [comment]
Have you solved the issue? Do you have a simple example for group_size >=2? I want to create a ranking model for ranking students by their achievements (in each class), where each class contains 20 students.
